docs/rfc/012-record-transaction-model.md

```diff
@@ -0,0 +1,243 @@
+```
+Yet another approach is to group interface functions inside a new block construct, e.g.
+```
+interface {
+
+    function mint(...) -> ... { ... }
+
+    function transfer(...) -> ... { ... }
+}
+```
+
+Now let us turn to the issue of clarifying how the Leo code reads old records and writes new records.
+
+Given that records have a fixed structure with typed slots,
+their format could be described by a Leo circuit type,
+whose member variables correspond to the slots.
+The types of the slots would be fairly low-level,
+i.e. byte arrays (e.g. `u8[128]` for the payload)
+and unsigned integers (e.g. `u64` for the balance),
+because they must have a clear correspondence with the serialized form of records.
+This means that the Leo code may have to do
+its own deserialization of the payload bytes into higher-level Leo values;
+standard serialization/deserialization libraries for Leo types may be provided for this,
+as an independent and more generally useful feature.
+
+It may make sense to have a circuit type for the special `input` variable,
+which includes two slots for the two old records.
+All these circuit types should be explicitly documented,
+and available to the Leo program.
+
+However, we probably want `input` to be read-only,
+i.e. disallow assigning to an old record slot.
+Designating `input` as `const` does not seem right,
+as that designation normally means that it is compiled into the circuit.
+Instead, we could provide read-only access via member function (e.g. `get_payload()`, `get_balance()`),
+but we still have to prohibit assignments to member variables (which is currently allowed on any circuit type).
+As an orthogonal and more generally useful feature,
+we could consider adding public/private access designations to Leo circuit members.
+Another approach is to avoid exposing the member variables,
+and just make the member functions available via an implicit import declaration.
+All of this needs to be thought through more carefully, in the broader context of the Leo language design;
+in any case, it should be clear that this can be made to work in some way,
+and that Leo programs can access the old records through the special `input` variables.
+
+One issue with the special `input` variable is whether it should be treated as a built-in global variable,
+or whether it should be explicitly passed to the interface functions and to the non-interface functions called by them.
+The first approach is more concise, while the second approach is more explicit.
+Note that, in the second approach, we may want to enforce certain restrictions on the use of `input`,
+e.g. we may not want to allow a call `f(input, input)` even if the parameters of `f` both have the same circuit type as `input`.
+That is, even if we make the passing of `input` explicit,
+there is still something special about it;
+in essence, it still needs to be treated essentially like a global variable,
+i.e. in a single-threaded way.
```

Comment:
This is not clear to me.  If input is immutable, or if it is passed by value, what is wrong with `f(input, input)`?

Reply:
I think you are right that there is probably nothing wrong per se since it's immutable, but it seems weird to me to do something like `f(input, input)`, also considering that `input` is a largish structure (each record in it has a 128-byte payload). It would be also nice to have a symmetric treatment with `output`. But again, I don't have strong preferences on any of this, so long as things are correct and clear.

Reply:
Not sure if immutable reference types make sense in Aleo/Leo but if so, that would be a solution to the 128-byte payload problem.

Reply:
Currently Leo does not have explicit reference types, but circuit values are implicitly passed by (mutable) reference to circuit member functions with `mut self` parameter. We could of course extend the type system; just mentioning the current situation.

Reply:
Perhaps we could pass `input` by *immutable* reference implicitly behind the scenes?

Reply:
Yes, we could. On the other hand, as @bendyarm pointed out, and if we go the more "functional" route of explicit `TransactionInput` and `TransactionOutput` parameter and result, there's no harm in copying the transaction input, or at least no more harm that copying a large user-defined structure. We've talked about having the compiler giving warnings when the generated R1CS constraints gets large (for some notion of 'large'), which may happen for a number of reasons. Maybe we could do that for passing large structures by value.